### PR TITLE
DAOS-7174 cont: take uuid as output param in cont_create

### PR DESCRIPTION
**Pull request by johannlombardi**
Change daos_cont_create() to take a uuid as output parameter.
The UUID cannot be provided by the user any longer and will
be returned to the caller.

Backward API compatibility is supported via a macro that use
the old behavior when a uuid_t is passed instead of a uuid_t *.

Backward ABI compatibility is supported by keeping around
the daos_cont_create symbols, but not advertising them via
the header files (the macro overrides it).

Rename daos_cont_create_by_label() to daos_cont_create_with_label()
to align with new function naming.

Signed-off-by: Johann Lombardi <johann.lombardi@intel.com>
Co-authored-by: Mohamad Chaarawi <mohamad.chaarawi@intel.com>

### CODE REVIEW

**Comment by johannlombardi** [comment]
Alternative simpler proposal. Just a quick sketch patch that likely does not compile yet just to share the idea. Anyway, i am really off now and leave it to you guys :)

**Comment by mchaarawi** [comment]
OK i think this is in good shape now to do reviews. Johann started this, but i updated it so i shouldn't review this myself.

File: src/include/daos_cont.h [hunk 1]
```diff
@@ -712,11 +723,24 @@ daos_cont_destroy_cpp(daos_handle_t poh, const uuid_t cont, int force, daos_even
 	uuid_unparse(cont, str);
 	return daos_cont_destroy2(poh, str, force, ev);
 }
+
+#define daos_cont_create daos_cont_create_cpp
+static inline int
+daos_cont_create_cpp(daos_handle_t poh, uuid_t *uuid, daos_prop_t *cont_prop, daos_event_t *ev)
+{
```
Comment (jolivier23) [inline src/include/daos_cont.h:730]: Please add a case for daos_cont_create to src/client/api/compile_check.c just to ensure it works for c++ now and into the future.

**Review by jolivier23 — CHANGES_REQUESTED** [review]

**Review by kccain — CHANGES_REQUESTED** [review]

File: src/client/dfs/dfs.c [hunk 2]
```diff
@@ -1352,15 +1354,21 @@ dfs_cont_create(daos_handle_t poh, uuid_t co_uuid, dfs_attr_t *attr,
...
+		rc = daos_cont_create(poh, cuuid, prop, NULL);
 	/* should not be freed by daos_prop_free */
 	prop->dpp_entries[prop->dpp_nr - 2].dpe_val_ptr = NULL;
 	if (rc) {
 		D_ERROR("daos_cont_create() failed "DF_RC"\n", DP_RC(rc));
 		D_GOTO(err_prop, rc = daos_der2errno(rc));
 	}
 
-	rc = daos_cont_open(poh, co_uuid, DAOS_COO_RW, &coh, &co_info, NULL);
+	if (uuid_is_set)
```
[the remark this answers is not in the document]
Comment (mchaarawi) [reply src/client/dfs/dfs.c]: good point

File: src/client/api/container.c [hunk 3]
```diff
@@ -87,16 +131,12 @@ daos_cont_create_by_label(daos_handle_t poh, const char *label,
...
 	}
 
-	uuid_generate(cuuid);
-	rc = daos_cont_create(poh, cuuid,
-			      merged_props ? merged_props : label_prop, ev);
+	rc = daos_cont_create2(poh, uuid, merged_props ? merged_props : label_prop, ev);
 	if (rc != 0) {
-		D_ERROR("daos_cont_create UUID:"DF_UUIDF" label=%s failed, "
-			DF_RC"\n", DP_UUID(cuuid), label, DP_RC(rc));
+		D_ERROR("daos_cont_create UUID: label=%s failed, "
```
[the remark this answers is not in the document]
Comment (mchaarawi) [reply src/client/api/container.c]: fixed

File: src/client/dfs/dfs.c [hunk 4]
```diff
@@ -1430,13 +1438,55 @@ dfs_cont_create(daos_handle_t poh, uuid_t co_uuid, dfs_attr_t *attr,
...
+		 dfs_t **dfs)
+{
+	uuid_t *u = (uuid_t *)((unsigned char *)cuuid);
+
+	return dfs_cont_create(poh, u, attr, coh, dfs);
+}
+
+/*
+ * Real latest & greatest implementation of container create. Used by anyone including the
+ * daos_cont.h header file.
```
[the remark this answers is not in the document]
Comment (mchaarawi) [reply src/client/dfs/dfs.c]: fixed

**Review by kccain — APPROVED** [review]

**Review by jolivier23 — APPROVED** [review]

**Pull request merged**